### ISSUE TITLE
Terraform hangs on plan/apply

### ISSUE THREAD

**Issue by LeftyBC** [hashicorp/terraform#2884]
Having an issue with Terraform hanging on apply or plan.  Setting TF_LOG=1 doesn't seem to reveal any useful information, and running strace isn't overly helpful as well.

Will attach my debug log.

**Comment by LeftyBC** [comment]
Debug log at: http://pastebitch.com/TOMO

**Comment by LeftyBC** [comment]
I've tried this with GOMAXPROCS set to 16, and unset.  

I've only recently fiddled with GOMAXPROCS (approximately the same time as this issue started) because I'm running a local Consul cluster for testing terraform -> Consul.

**Comment by ryanking** [comment]
I'm having the same problem since yesterday:

https://gist.github.com/ryanking/94986ce2450984d77358

When it stalls there are typically a number of processes running:

```
ryanking        13346   0.3  0.1 145171076   5208 s000  S+   10:32AM   0:00.02 /Users/ryanking/bin/terraform-provisioner-remote-exec
ryanking        13347   0.3  0.1 145171092   5192 s000  S+   10:32AM   0:00.02 /Users/ryanking/bin/terraform-provisioner-file
ryanking        13345   0.2  0.1 145169868   4800 s000  S+   10:32AM   0:00.02 /Users/ryanking/bin/terraform-provisioner-local-exec
ryanking        13348   0.2  0.1 145171112   5312 s000  S+   10:32AM   0:00.03 /Users/ryanking/bin/terraform-provisioner-chef
ryanking        13350   0.0  0.1 145181524   9936 s000  S+   10:32AM   0:00.09 /Users/ryanking/bin/terraform-provider-aws
ryanking        13349   0.0  0.1 145170208   4792 s000  S+   10:32AM   0:00.03 /Users/ryanking/bin/terraform-provider-cloudflare
ryanking        13342   0.0  0.4 145174588  31296 s000  S+   10:32AM   0:00.81 /Users/ryanking/bin/terraform plan
ryanking        13341   0.0  0.1 145174332   5256 s000  S+   10:32AM   0:00.19 terraform plan
```
Running dtruss on them reveals a lot of select looping:

```
...
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
gettimeofday(0x208C5CED0, 0x0, 0x0)		 = 1438190792 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
select(0x0, 0x0, 0x0, 0x0, 0x208C55F38)		 = 0 0
...
```

I first had the problem on 0.6.0 yesterday. Upgrading to 0.6.1 didn't help. I tried older versions of my terraform configs and those also hung.

A colleague did a run on 0.5.3 and saw the same problem.

I've tried setting GOMAXPROCS to a ridiculously high number and that didn't help.

**Comment by phinze** [comment]
This is definitely odd. I'm suspicious of the "close" nodes I'm seeing towards the bottom of both of your debug logs...

From the 0.6.0 changelog:

> core: Properly close provider and provisioner plugin connections [GH-2406, GH-2527]

I know one of you had a coworker try v0.5.3... can you both double check with v0.5.3?

**Comment by ryanking** [comment]
Here is my coworker running the same configs with 0.5.3:

https://gist.github.com/anonymous/e27884ec1d834ae16b33

it stalls too

**Comment by LeftyBC** [comment]
Debug log for 0.5.3: http://pastebitch.com/CATC

I did have to change the catalog a bit to get around some 0.5.x bugs with providers, so it's not identical, but it still stalls.

**Comment by LeftyBC** [comment]
Running it under strace with just network tracing on shows Terraform trying to connect to a 169.254.169.254 address, I wonder if it has something to do an AWS provider?

http://pastebitch.com/eZAJ

**Comment by ryanking** [comment]
@LeftyBC Interesting that you have several processes trying to connect there.

**Comment by ryanking** [comment]
Perhaps this timeout is not working:

https://github.com/hashicorp/terraform/blob/c72173853d35f4c11e751c95a1a44c0b16e915a3/builtin/providers/aws/provider.go#L39

**Comment by phinze** [comment]
These are good theories. Trying to do a bit of differential diagnosis here.

The metadata service `169.254.169.254` authentication mechanism was added between v0.6.0 and v0.6.1 https://github.com/hashicorp/terraform/commit/da1cac623d9ebf96355bb28c003bec0887cc3191

The close provider feature was added between v0.5.3 and v0.6.0.

So the fact that it stalls for both of you on 0.5.3 as well seems to discredit those theories.

Does any binary searching through your configs by commenting out portions seem to affect the reproduction of the hanging?

**Comment by phinze** [comment]
Similarly, do any runs from your commit history avoid the hang? Hopefully with more data like this we can narrow down the issue.

**Comment by LeftyBC** [comment]
I'll go back through the commit log and see if I can get it to not hang, stand by.

**Comment by LeftyBC** [comment]
Going back several versions in my commit log doesn't seem to help, even commits that I *know* were working a week ago, etc.  Trying with 0.5.3 and 0.6.2 each time I go back a commit still gives me the hang.

Debug log for 0.6.1 with GOMAXPROCS=254 on a commit that I know should work: http://pastebitch.com/DAfA

Here's the strace output for the run that generated the above debug log - I killed it just after the normal hang:  http://pastebitch.com/eFbg

**Comment by ryanking** [comment]
Commenting things out doesn't help, because it still reads the state file and has to consider all the resources.

I commented out everything and it still hangs.

**Comment by LeftyBC** [comment]
I'll try moving the state file away temporarily.

**Comment by LeftyBC** [comment]
Moving the state file stops it hanging... not sure why I didn't think of that in the first place.  Of course, that means we lose all the resources that Terraform is managing, but that looks to me like it's just a state corruption inssue.

@phinze If you need me to post a (redacted, naturally) copy of our state file, let me know if that would help.

**Comment by ryanking** [comment]
Ah ha. Same for me.

**Comment by ryanking** [comment]
I got it to work! The state file gave me the clue to figuring out how to work around this.

I have been working recently to refactor our setup into modules so that it can be more reusable. In the process my state file accumulated a lot of empty module state. For example, from `terraform show`:
```
...
module.test.vpc
  0 resource(s)
...
```

I went through the state file and manually removed a bunch of these empty modules. When I did, suddenly things started working again.

**Comment by LeftyBC** [comment]
I too have been refactoring to use modules, so that makes some sense. For me deleting the empty modules doesn't seem to have solved it, though.

For now I'm going to archive the state file and repo exactly how it is, then delete the state file and terraform-created resources and let everything rebuild.

**Comment by phinze** [comment]
Great debugging work, @ryanking and @LeftyBC!

Empty modules in the state is a great find. We can definitely investigate that path.

**Comment by LeftyBC** [comment]
Not sure how easy it will be to repro from your side, but on my end I created a module that didn't have any resources in it, just variables for doing lookups on (VPC and subnet IDs for our different environments).  I initally referenced this module from the root, but then moved it into an "environment" submodule.  It was referencing the same module at the same path, just from different levels of submodules.

Inital structure went root -> vpc_module
Changed to root -> environment_module -> vpc_module

Hope that helps.

**Comment by phinze** [comment]
@LeftyBC after moving it to an "environment" submodule, that's when things started hanging?

**Comment by LeftyBC** [comment]
I think so.  I also changed the name of the resource that was referring to the module after I moved it to the environment submodule.

In the root, it was 
```
module "cts_vpc" {
    source = "modules/components/cts_vpc"
...
}
```

It was then moved to the environment submodule in `modules/environments/opsdev`:

```
module "cts_vpc" {
   source = "../../components/cts_vpc"
...
}
```

I then changed it to:
```
module "cts_nonprod_vpc" {
  source = "../../components/cts_vpc"
...
}
```

I believe this is what left a bunch of "empty" modules in the state, as @ryanking saw.

**Comment by phinze** [comment]
Hey folks - this should be fixed in Terraform 0.6.3. Let me know if you still have any troubles on that version!

**Comment by typerlc** [comment]
I have this issue with 0.6.3 on Mac OSX Yosemite.  In case it matters, it was installed via brew.

Running terraform plan on the first getting start guide example.tf causes a hang, with the following repeatedly output when TF_LOG=1

$ TF_LOG=1 terraform plan

2015/08/28 12:34:45 [DEBUG] vertex provider.aws (close), waiting for: aws_instance.example
2015/08/28 12:34:45 [DEBUG] vertex aws_instance.example, waiting for: provider.aws
2015/08/28 12:34:45 [DEBUG] vertex root, waiting for: provider.aws (close)

$ terraform show
No state.

**Comment by phinze** [comment]
@typerlc weird! It's possible that could happen due to transient networking issues - is this something that happens every time for you? If so, can you open a fresh issue with the config that reproduces? Thanks!

**Comment by apparentlymart** [comment]
@typerlc is it possible that you ran ``terraform plan`` without AWS credentials set, either in your config or in the environment?

The AWS provider has some special behavior to try to obtain instance role credentials as if it were running on an EC2 instance, which applies if it doesn't find any explicit credentials in the configuration. This involves reaching out to ``http://169.254.169.254/latest/meta-data/``, which of course won't do the right thing if you're not actually running on an EC2 instance... it could do anything from time out trying to send packets into oblivion, to getting a "no route to host" error, to actually talking to a different machine in your local network that happens to have that IP address.

I've personally run into this a few times and it ends up *looking* a lot like a hang. I expect it could be addressed by setting a pretty aggressive timeout on that request, assuming that the metadata endpoint on EC2 should respond reasonably quickly if you're actually on an EC2 instance.

**Comment by typerlc** [comment]
@phinze I doubt its a networking issue.  Happens from my office network, mobile phone network and home network.  While doing this, I have also browsed the internet, accessed the AWS console, etc.
As requested, I've opened a new issue #3118 to track.

@apparentlymart My .tf file had my AWS credentials.  I've included my config file in the new issue #3118.

**Comment by trevorr** [comment]
For anyone hitting this issue against AWS, be sure to check the [AWS Service Health Dashboard](http://status.aws.amazon.com/). For instance, currently IAM N. Virginia is reporting `We are continuing to investigate increased error rates and latencies on IAM administrative APIs and Console.` This can cause `aws.iam.GetUser` to take several minutes to complete (possibly retrying multiple times after getting a 503 response).
